utils/src/main/java/software/amazon/awssdk/utils/async/DelegatingBufferingSubscriber.java

```diff
@@ -0,0 +1,137 @@
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.ON_NEXT;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * This Subscriber buffers the {@link ByteBuffer} it receives until the specififed maxiimumBuffer is reached, at which point
+ * it sends those bytes to the delegate {@link Subscriber}.
+ */
+@SdkInternalApi
+public class DelegatingBufferingSubscriber extends DelegatingSubscriber<ByteBuffer, ByteBuffer> {
+
+    /**
+     * The maximum amount of bytes allowed to be stored in the StoringSubscriber
+     */
+    private final int maximumBuffer;
+
+    /**
+     * Current amount of bytes buffered in the StoringSubscriber
+     */
+    private final AtomicLong currentlyBuffered = new AtomicLong(0);
+
+    /**
+     * Stores the bytes received from the upstream publisher, awaiting sending them to the delegate once the buffer size is
+     * reached.
+     */
+    private final StoringSubscriber<ByteBuffer> storage = new StoringSubscriber<>(Integer.MAX_VALUE);
+
+    /**
+     * Publisher to send the ByteBuffer received from upstream publisher to the StoringSubscriber
+     */
+    private final SimplePublisher<ByteBuffer> publisherToStorage = new SimplePublisher<>();
+
+    /**
+     * Subscription of the publisher this subscriber subscribe to
+     */
+    private Subscription subscription;
+
+    public DelegatingBufferingSubscriber(int maximumBuffer, Subscriber<? super ByteBuffer> subscriber) {
+        super(subscriber);
+        this.maximumBuffer = maximumBuffer;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        super.onSubscribe(subscription);
+        this.subscription = subscription;
+        publisherToStorage.subscribe(storage);
+        subscription.request(maximumBuffer);
+    }
+
+    @Override
+    public void onNext(ByteBuffer byteBuffer) {
+        // First, if the incoming ByteBuffer would make the total buffered amount exceed the maximum buffer, send what is
+        // currently buffered in the storage to the delegate.
+        if (currentlyBuffered.get() + byteBuffer.remaining() > maximumBuffer) {
+            flushStorageToDelegate();
+        }
+
+        // Then, if the incoming ByteBuffer would still alone bust the buffer size, remove one chunk of it (of maximumBuffer
+        // size) then try again
+        if (currentlyBuffered.get() + byteBuffer.remaining() > maximumBuffer) {
```

Comment:
This is the same condition as in the clause above.

Reply:
yes, but `currentlyBuffered.get()` and `byteBuffer.remaining()` would have changed

Reply:
It is needed in the case where the ByteBuffer is itself larger than the maximumBuffer, we need to send chunks of it to the delegate